x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/common/AbstractCompositeAggFunction.java

```diff
@@ -88,9 +88,12 @@ public void preview(
                         return;
                     }
                     final CompositeAggregation agg = aggregations.get(COMPOSITE_AGGREGATION_NAME);
+                    if(agg == null){
+                        return;
```

Comment:
the `return` is ok, however the passed `listener` in this method has to be called _exactly once_, either using `onFailure` or `onResponse`.

In this case the composite agg returns nothing, because there is no data that fits, this is not an error, therefore you should call

```
listener.onResponse(Collections.emptyList());
```

I would also add the empty case:

```
 if (agg == null || agg.getBuckets().isEmpty()) 
 ```

Reply:
Thanks for pointing this out, I missed the listener response. I have added the additional condition as you pointed out as well, I thought about this before, since it didnt cause any issue so I left it but definitely a good pruning.